# Day 1

In [2]:
print('Hello Raghu')

Hello Raghu


In [3]:
!date

Mon 07 Feb 2022 04:43:44 AM UTC


In [5]:
!ls

day1.ipynb  day2.ipynb	day3.ipynb  day4.ipynb	day5.ipynb


In [6]:
!python --version

Python 3.8.5


In [8]:
%%file hello.py
print('hello')

Writing hello.py


In [10]:
!cat hello.py

print('hello')


In [11]:
!python hello.py

hello


#### Data Types

Python has integers.

In [12]:
2+4

6

In [13]:
2**100

1267650600228229401496703205376

In [15]:
.1+.2

0.30000000000000004

In [16]:
.2+.7

0.8999999999999999

In [17]:
.2+.4

0.6000000000000001

In [26]:
print('Raghu\n'*4)

Raghu
Raghu
Raghu
Raghu



In [18]:
print('a\nb\nc\n')

a
b
c



In [128]:
print('Name\tAge\tCountry') # \t is tab space 
print('Raghu\t45\tIndia')

Name	Age	Country
Raghu	45	India


In [115]:
print('='*20)

In [21]:
message = """ 
Hello Raghu, 
Welcome to Advanced Python Training 
"""

In [22]:
message

' \nHello Raghu, \nWelcome to Advanced Python Training \n'

In [23]:
print(message)

 
Hello Raghu, 
Welcome to Advanced Python Training 



#### Python string support unicode as well

In [119]:
print("అ ఆ ఇ ఈ") # Telugu
print("ಅ ಆ ಇ ಈ") # Kannada
print("அ ஆ இ ஈ") # Tamil
print("അ ആ ഇ ഈ") # Malayalam
print("अ आ इ ई") # Hindi

అ ఆ ఇ ఈ
ಅ ಆ ಇ ಈ
அ ஆ இ ஈ
അ ആ ഇ ഈ
अ आ इ ई


In [120]:
print("\u0c05 \u0c06 \u0c07 \u0c08") # Telugu
print("\u0c85 \u0c86 \u0c87 \u0c88") # Kannada
print("\u0b85 \u0b86 \u0b87 \u0b88") # Tamil
print("\u0d05 \u0d06 \u0d07 \u0d08") # Malayalam
print("\u0905 \u0906 \u0907 \u0908") # Hindi

అ ఆ ఇ ఈ
ಅ ಆ ಇ ಈ
அ ஆ இ ஈ
അ ആ ഇ ഈ
अ आ इ ई


In [39]:
print("\U0001f639")

😹


In [40]:
print("\U0001f638")

😸


In [43]:
print("\U0001f645")

🙅


In [44]:
print("\U0001f648")

🙈


#### Bytes

Python has a `bytes` data type to represent binary data

bytes are written just like string with a b prefix.

In [122]:
data = b'12234'

In [124]:
type(data)

bytes

#### Lists

In [35]:
x = ['a','b',[1,3,5]]

In [36]:
x[0]

'a'

In [37]:
x[2]

[1, 3, 5]

In [ ]:
len(x)

#### Dictionaries

In [130]:
person = {
    'name':'Raj',
    'email':'raju@gamil.com',
    'phone':123456,
    'country':'India'
}

In [131]:
person['name']

'Raj'

In [133]:
person['country']

'India'

#### Boolean Values

In [135]:
True

True

In [136]:
False

False

In [137]:
filesize = 20

In [138]:
filesize>10

True

In [139]:
filesize >30

False

#### Other Data types

Python has a `None` datatypes to represent nothing

In [141]:
x = None

In [142]:
print(x)

None


##### Python has tuple data types

In [143]:
nums = (2,4,6,8,10)

In [144]:
print(nums,type(nums))

(2, 4, 6, 8, 10) <class 'tuple'>


In [145]:
nums[0]

2

In [146]:
nums[-2]

8

In [147]:
a,b,c,d,e = nums

In [148]:
a

2

In [149]:
d

8

In [150]:
yellow = 255,255,0 # r,g,b

In [151]:
yellow

(255, 255, 0)

Python also has a `set` datatypes. A set is an unordered collection of unique elements

In [152]:
{1,2,3,5,7,8,3,2,5,8}

{1, 2, 3, 5, 7, 8}

#### Functions

Python has many built-in functions

In [153]:
print('Raghu')

Raghu


In [154]:
len('Raghu')

5

In [155]:
type('Raghu')

str

You may have noticed that Python doesn't allow operations on incompatible data types.

In [156]:
1 + '2'

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [157]:
1 + int('2')

3

In [158]:
str(1) + '2'

'12'

**Example: Counting the number of digits in a number**

*How to count the number of digits in a number?*

In [160]:
2**100

1267650600228229401496703205376

In [162]:
2**1000

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

In [164]:
n = 2**1000
print(len(str(n)))

302


writing our own function

In [166]:
def square(x):
    return x*x
square(5)

25

Problem: Write a cube function to compute cube of a number.

    >>> cube(2)
        8
    >>> cube(3)
        27
Problem: Write a function count_digits that takes a number as argument and returns the number of digits it has.

    >>> count_digits(12345)
        5
    >>> count_digits(2**1000)
        302

In [167]:
def cube(y):
    return y**3
cube(5)

125

In [168]:
def count_digits(n):
    return len(str(n))
count_digits(362746273647)

12

In [169]:
count_digits(1514)

4

**print vs return**

In [170]:
def square1(x):
    return x*x

def square2(x):
    print(x*x)

In [171]:
square1(5)

25

In [173]:
square2(5)

25


What is the difference between these two? Which one is better?

Can you suggest some code that works with one, but won't with the other.

Let's try to compute x^2 +1

In [174]:
square1(4) + 1

17

It is not possible to compute x^2+1 using square2 because it is just printing the result of x^2, but not giving that value back to us to use in other computations.

In [176]:
square2(4) + 1

16


TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'

**Functions are values too!**

In [177]:
def square(x):
    return x*x

In [178]:
square(4)

16

In [179]:
square

<function __main__.square(x)>

In [180]:
f = square

In [181]:
f(4)

16

In [182]:
f

<function __main__.square(x)>

In [183]:
def square(x):
    return x*x

# square = make-new-function..

In [184]:
def sum_of_squares(x, y):
    return square(x) + square(y)

In [185]:
sum_of_squares(3, 4)

25

**What if we want to compute sum of cubes as well ?**

In [188]:
def cube(x):
    return x*x*x

def sum_of_cubes(x, y):
    return cube(x) + cube(y)

In [190]:
sum_of_cubes(3, 4)

91

**Can generalize sum_of_squares and sum_of_cubes into a single function?
yes, By passing the function as argument in another function**

In [192]:
def sum_of(f,x,y):
    return f(x)+f(y)

In [194]:
sum_of(square,2,6)

40

In [195]:
sum_of(cube,2,6)

224

In [197]:
sum_of(len, "hello", "python")

11

In [199]:
sum_of(abs, 4, -10)

14

**The feature of passing functions as arguments is so useful that many built-in functions support that.**

In [200]:
max(122,125)

125

In [201]:
max([3, 5, 4])

5

In [203]:
#by default it give max value by dictionary basis
max(['one', 'two', 'three', 'four', 'five'])

'two'

How to find the longest word?

In [209]:
#background it take each element , find its length then give max length letter
# key is used to assign a built_in or normal fucntion
max(['one', 'two', 'three', 'four', 'five'], key=len)

'three'

In [207]:
#the bellow function run in background for the above statement
def mylen(word):
    print("mylen", word)
    return len(word)

In [210]:
max(['one', 'two', 'three', 'four', 'five'], key=mylen)

mylen one
mylen two
mylen three
mylen four
mylen five


'three'

In [211]:
# records of students with marks
records = [
    ("A", 78),
    ("B", 91),
    ("C", 56)
] 

**How to find the student who got the maximum marks?**

In [212]:
def get_marks(record):
    return record[1]

max(records, key=get_marks)

('B', 91)

In [213]:
max(records, key=get_marks)[0]

'B'

In [214]:
h1 = {"name": "web-0", "type": "web", "loadavg": 67}
h2 = {"name": "web-1", "type": "web", "loadavg": 78}
h3 = {"name": "db-0", "type": "dv", "loadavg": 42}

In [215]:
hosts = [h1, h2, h3]

In [216]:
hosts

[{'name': 'web-0', 'type': 'web', 'loadavg': 67},
 {'name': 'web-1', 'type': 'web', 'loadavg': 78},
 {'name': 'db-0', 'type': 'dv', 'loadavg': 42}]

In [217]:
def get_loadavg(host):
    return host['loadavg']

max(hosts, key=get_loadavg)

{'name': 'web-1', 'type': 'web', 'loadavg': 78}

In [218]:
max(hosts, key=get_loadavg)['name']

'web-1'

Methods

In [219]:
x = "Hello"

In [220]:
x.upper()

'HELLO'

upper is a method available to all strings.

In [221]:
y=1
# this doesn't work as int don't support upper method
y.upper()

AttributeError: 'int' object has no attribute 'upper'

In [222]:
x.lower()

'hello'

In [223]:
"mathematics".count("mat")

2

In [224]:
"mathematics".replace("mat", "rat")

'ratheratics'

Problem: Write a function count_zeros that counts the number of zeros in the given number.

    >>> count_zeros(0)
        1
    >>> count_zeros(100)
        2
    >>> count_zeros(2030405)
        3

In [225]:
def count_zeros(n):
    return str(n).count('0')

In [226]:
count_zero(10)

1

In [227]:
count_zero(100)

2

In [228]:
count_zeros(2030405)

3

**Let's look at some more intersting methods on strings.**

In [229]:
sentence = "Anything that can go wrong, will go wrong."

In [230]:
sentence.split() # split on any whitespace

['Anything', 'that', 'can', 'go', 'wrong,', 'will', 'go', 'wrong.']

In [231]:
sentence.split(",") # split on comma

['Anything that can go wrong', ' will go wrong.']

In [232]:
sentence.split("go") # split on the word go

['Anything that can ', ' wrong, will ', ' wrong.']

The `split` method splits a string based on any delimiter. If no delimiter is provided, it uses any whitespace as the delimiter.

What is the reverse of split?

In [233]:
"-".join(["a", "b", "c"])

'a-b-c'

In [234]:
" ".join(["a", "b", "c"])

'a b c'

In [235]:
"::".join(["a", "b", "c"])

'a::b::c'

Problem: Write a function `longest_word` that takes a sentence as argument and returns ther longest word from it.

    >>> longest_word("one two three four five")
            'three'
    >>> longest_word("Anything that can go wrong, will go wrong.")
            'Anything'

In [236]:
sentence = "one two three four five"

In [237]:
sentence.split()

['one', 'two', 'three', 'four', 'five']

In [238]:
words = sentence.split()

In [239]:
words

['one', 'two', 'three', 'four', 'five']

In [240]:
def longest_word(sentence):
    words = sentence.split()    
    return max(words, key=len)

In [241]:
longest_word("one two three four five")

'three'

**Modules**

**Time module**

In [242]:
import time

In [243]:
time.asctime()

'Mon Feb  7 18:02:23 2022'

In [244]:
!date

Mon 07 Feb 2022 06:02:34 PM UTC


In [245]:
%%file date.py
import time
print(time.asctime())

Writing date.py


In [246]:
!python date.py

Mon Feb  7 18:03:04 2022


**OS module**

In [249]:
import os
# get current #working directory
os.getcwd()

'/home/nraghu'

In [250]:
# list all files in a directory
os.listdir(".")

['square.py',
 'day4.ipynb',
 'day5.ipynb',
 'hello.py',
 '.bash_logout',
 'day3.ipynb',
 '.ipynb_checkpoints',
 '.bashrc',
 'day1.ipynb',
 '.profile',
 'day2.ipynb',
 '.cache',
 '.jupyter',
 '.local',
 '.ipython',
 'date.py',
 '.juputer',
 'header.py']

If we want to find size of a file, we can use os.path.getsize.

In [252]:
os.path.getsize("day1.ipynb")

64229

In [253]:
!ls -l day1.ipynb

-rw-r--r-- 1 nraghu nraghu 62628 Feb  7 18:06 day1.ipynb


how to find the largest file in the current directory?

In [254]:
max(os.listdir("."), key=os.path.getsize)

'day1.ipynb'

In [255]:
!ls -lS

total 96
-rw-r--r-- 1 nraghu nraghu 63037 Feb  7 18:07 day1.ipynb
-rw-r--r-- 1 nraghu nraghu   337 Feb  6 21:00 day5.ipynb
-rw-r--r-- 1 nraghu nraghu   336 Feb  6 21:00 day2.ipynb
-rw-r--r-- 1 nraghu nraghu   335 Feb  6 21:00 day3.ipynb
-rw-r--r-- 1 nraghu nraghu   331 Feb  6 21:00 day4.ipynb
-rw-rw-r-- 1 nraghu nraghu    49 Feb  7 07:39 square.py
-rw-rw-r-- 1 nraghu nraghu    34 Feb  7 18:02 date.py
-rw-rw-r-- 1 nraghu nraghu    30 Feb  7 09:18 header.py
-rw-rw-r-- 1 nraghu nraghu    15 Feb  7 04:46 hello.py


**very very important**

**Reading Command-line arguments**

The `sys` module keeps track of all arguments passed to the program.

In [256]:
%%file args.py
import sys
print(sys.argv)

Writing args.py


In [257]:
!python args.py

['args.py']


In [258]:
!python args.py hello

['args.py', 'hello']


In [260]:
!python args.py hello workd

['args.py', 'hello', 'workd']


In [261]:
!python args.py 1 2 3 4 5

['args.py', '1', '2', '3', '4', '5']


Notice that the arguments are always strings.

Example: echo

In [262]:
!echo hello world

hello world


Let's implement it in Python. For now, we'll print only the first argument.

In [263]:
%%file echo.py
import sys
arg = sys.argv[1]
print(arg)

Writing echo.py


In [265]:
!python echo.py hello

hello


In [266]:
!python echo.py hello world

hello


Problem: Write a program square.py that takes a number as command-line argument and prints its square.

      >>> python square.py 2
        4
      >>> python square.py 5
        25

In [276]:
%%file square.py
import sys
arg = sys.argv[1]
print(int(arg)**2)

Overwriting square.py


In [277]:
!python square.py 5

25


In [278]:
%%file cube.py
import sys
arg = sys.argv[1]
print(int(arg)**3)

Overwriting cube.py


In [279]:
!python square.py 3

9


            Q & A

                              imax
1)Write a function imax to compute maximum of given words ignoring the case.

    >>> imax(['alice', 'Dave', 'charlie', 'BOB'])
            'Dave'

    >>> imax(['Alice', 'dave', 'charlie', 'BOB'])
            'dave'

In [305]:
def ignore_case(word):
    return word.lower()

def imax(l):
    return max(l,key=ignore_case)

In [306]:
print(imax(['alice', 'Dave', 'charlie', 'BOB']))

Dave


In [307]:
print(imax(['Alice', 'dave', 'charlie', 'BOB']))

dave



2)Write a function digit_count that takes a number and a digit as argument and returns the number of times the digit is present in that number.

    >>> digit_count(1231, 1)
            2
    >>> digit_count(1231, 3)
            1
    >>> digit_count(1231, 9)
            0

In [308]:
def digit_count(a,b):
    return (str(a).count(str(b)))

In [309]:
digit_count(1231, 1)

2

In [310]:
digit_count(1231, 3)

1

In [311]:
digit_count(1231, 9)

0

                            Make Header
3)Write a program header.py that takes a word as command-line argument and prints it as header as shown below with the word converted to upper case.

        $ python header.py python
                ======
                PYTHON
                ======

        $ python header.py python-foundation-course
                ========================
                PYTHON-FOUNDATION-COURSE
                ========================

        Hint:

        >>> name = 'python'
        >>> name.upper()
        'PYTHON'

In [312]:
# you need to pass the input from command prompt
import sys
s = sys.argv[1]
print('='*len(s))
print(s.upper())
print('='*len(s))

==
-F
==


                            Text in box
4)Write a program box.py that takes word as a command-line argument and prints the word in a box as shown below.

        $ python box.py python
            +--------+
            | python |
            +--------+
Please note that there should be exactly one space on either side of the text in the box.



In [313]:
# you need to pass the input from command prompt
import sys
ar = sys.argv[1]
length = len(ar)+2
print('+' + '-'*length + '+')
print(f'| {ar} |')
print('+' + '-'*length + '+')

+----+
| -f |
+----+


                            Add two numbers
5)Write a program add.py that takes two numbers as command-line arguments and prints their sum.

     >>> $ python add.py 3 4
            7
 >>> $ python add.py 10 20
            30

In [318]:
# you need to pass the input from command prompt
import sys
print(int(sys.argv[1])+int(sys.argv[2]))